numpy/random/mtrand/mtrand.pyx

```diff
@@ -4967,33 +4969,53 @@ cdef class RandomState:
                [0, 1, 2]])
 
         """
-        cdef npy_intp i, j
-
-        i = len(x) - 1
-
-        # Logic adapted from random.shuffle()
-        if isinstance(x, np.ndarray) and \
-           (x.ndim > 1 or x.dtype.fields is not None):
-            # For a multi-dimensional ndarray, indexing returns a view onto
-            # each row. So we can't just use ordinary assignment to swap the
-            # rows; we need a bounce buffer.
+        cdef:
+            npy_intp i, j, n = len(x)
+            size_t stride, nbytes
+            char* x_ptr
+            char* buf_ptr
+
+        if type(x) is np.ndarray and x.ndim == 1 and x.size:
+            # Fast, statically typed path: shuffle the underlying buffer.
+            # Only for non-empty, 1d objects of class ndarray (subclasses such
+            # as MaskedArrays may not support this approach).
+            x_ptr = <char*><size_t>x.ctypes.data
```

Comment:
Because Cython interprets `<char*>python_object` as casting a bytes object (*not* an object supporting the buffer protocol) to a `char*`, but `<char*>c_object` as a C-style cast.  (If you try it you'll get "TypeError: expected bytes, int (or whatever you pass in) found")